File: cli/src/core/TopicIndexStore.ts
/**
 * TopicIndexStore — read/write `topics/index.json`, the routing index for the
 * topic KB. Persisted via the active StorageProvider (dual-write).
 */

import { createLogger } from "../Logger.js";
import type { FileWrite } from "../Types.js";
import type { StorageProvider } from "./StorageProvider.js";
import { resolveStorage } from "./SummaryStore.js";
import type { TopicIndex } from "./TopicKBTypes.js";

const log = createLogger("TopicIndexStore");
const INDEX_PATH = "topics/index.json";

/** A fresh, empty index. */
export function emptyTopicIndex(): TopicIndex {
	return { schemaVersion: 1, topics: [] };
}

/** Reads `topics/index.json`; missing or unparseable → empty index (never throws). */
export async function readTopicIndex(cwd?: string, storage?: StorageProvider): Promise<TopicIndex> {
	const resolved = resolveStorage(storage, cwd);
	const raw = await resolved.readFile(INDEX_PATH);
	if (!raw) return emptyTopicIndex();
	try {
		const parsed = JSON.parse(raw) as TopicIndex;
		return { schemaVersion: 1, topics: parsed.topics ?? [] };
	} catch {
		log.warn("Failed to parse %s — treating as empty", INDEX_PATH);
		return emptyTopicIndex();
	}
}

/** Persists the index via the active StorageProvider. */
export async function saveTopicIndex(index: TopicIndex, cwd?: string, storage?: StorageProvider): Promise<void> {
	const resolved = resolveStorage(storage, cwd);
	const files: FileWrite[] = [{ path: INDEX_PATH, content: JSON.stringify(index, null, "\t") }];
	await resolved.writeFiles(files, `Update topic KB index (${index.topics.length} topics)`);
}

File: intellij/src/main/kotlin/ai/jolli/jollimemory/core/TopicIndexStore.kt
package ai.jolli.jollimemory.core

/**
 * TopicIndexStore — read/write `topics/index.json`, the routing index for the
 * topic KB, via a [StorageProvider].
 *
 * Kotlin port of `cli/src/core/TopicIndexStore.ts`.
 */
object TopicIndexStore {

    private val log = JmLogger.create("TopicIndexStore")
    private const val INDEX_PATH = "topics/index.json"

    /** A fresh, empty index. */
    fun emptyTopicIndex(): TopicIndex = TopicIndex(schemaVersion = 1, topics = emptyList())

    /** Reads `topics/index.json`; missing or unparseable → empty index (never throws). */
    fun readTopicIndex(storage: StorageProvider): TopicIndex {
        val raw = storage.readFile(INDEX_PATH) ?: return emptyTopicIndex()
        val parsed = TopicJson.parse(raw, TopicIndex::class.java)
        if (parsed == null) {
            log.warn("Failed to parse %s — treating as empty", INDEX_PATH)
            return emptyTopicIndex()
        }
        // Normalize like the TS reader (parsed.topics ?? []): Gson injects null into the
        // non-null `topics` when the key is absent/null, so coalesce explicitly.
        @Suppress("USELESS_ELVIS")
        val topics = parsed.topics ?: emptyList()
        return TopicIndex(schemaVersion = 1, topics = topics)
    }

    /** Persists the index via the provider. */
    fun saveTopicIndex(index: TopicIndex, storage: StorageProvider) {
        storage.writeFiles(
            listOf(FileWrite(path = INDEX_PATH, content = TopicJson.stringify(index))),
            "Update topic KB index (${index.topics.size} topics)",
        )
    }
}
